spanconfigkvsubscriber: teach the KVSubscriber about system span configs

arulajmani opened the pull request:
See individual commits for details.

arulajmani committed:
spanconfigkvsubscriber: improve test

Simplify test and unify file name with others in the package.

Release note: None

arulajmani committed:
spanconfigkvsubscriber: add test for decoding system span config targets

This worked but didn't have a test previously.

Release note: None

arulajmani committed:
spanconfigkvsubscriber: teach the KVSubscriber about system span configs

This patch teaches the KVSubscriber to handle updates to system span
configurations. There's not much here, other than invoking registered
handlers with the correct span when a system span config is updated.
This entails:
- The everything span when the system span config targets the entire
keyspace or targets the host tenant.
- The targeted tenant's span if the system span config targets a
secondary tenant.

Release note: None

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go:
```diff
@@ -86,6 +77,83 @@ func TestSpanConfigDecoder(t *testing.T) {
+	}
+
+	for i, systemTarget := range []spanconfig.SystemTarget{
+		// Tenant targeting its logical cluster.
```
Drop the "cluster" verbiage here and throughout.

irfansharif commented on pkg/spanconfig/systemtarget.go:
```diff
@@ -149,6 +149,27 @@ func (st SystemTarget) targetsEntireKeyspace() bool {
+		return keys.EverythingSpan
+	}
+	switch st.systemTargetType {
+	case SystemTargetTypeEntireKeyspace:
```
Why do we need both this and `st.targetsEntireKeyspace` above?

irfansharif commented on pkg/spanconfig/systemtarget.go:
```diff
@@ -149,6 +149,27 @@ func (st SystemTarget) targetsEntireKeyspace() bool {
+			EndKey: k.PrefixEnd(),
+		}
+	case SystemTargetTypeAllTenantKeyspaceTargetsSet:
+		panic("read only target")
```
This panic feels a bit misplaced/confusing. Not apparent that we're doing any writes here in this API.

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go:
```diff
@@ -46,8 +46,9 @@ import (
 //      delete [c,e)
 //      upsert [c,d):C
 //      upsert [d,e):D
+//      upsert {entire-keyspace}:_CLUSTER
```
Why the _? Drop the cluster verbiage. The text here is just a shorthand for a config, using any letter in the alphabet instead would avoid confusing readers that these words/underscore have special meaning. Mind doing that?

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go:
```diff
@@ -243,7 +245,7 @@ func TestDataDriven(t *testing.T) {
 
 					var spanStr string
-					if update.Equal(keys.EverythingSpan) {
+					if update.Equal(keys.EverythingSpan) || update.Key.Compare(keys.TenantTableDataMin) > 0 {
```
Why did we need this?

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+
+store-reader key=e
+----
+conf=D_CLUSTER
```
Instead of concatenating and making use of _'s, what if this syntax captured exactly what was being appended? 
```
conf=D+X
```

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+# Update span configs that target tenant keyspaces and ensure handlers are
+# invoked correctly.
+update
+upsert {source=1, target=1}:_HOST
```
[nit] Here too I'd prefer using single letters instead and showing combination of configs through "A+B+C".

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+upsert {source=10, target=10}:_TEN
+----
+
+# M{in-ax} corresponds to the system tenant targeting itself.
```
Is this right? I thought the system tenant targeting itself would would only capture its own keyspace without going into the tenant's. Wasn't that the whole point of separating out the EntireKeyspaceTarget into its own thing?

arulajmani reviewed (commented):
RFAL

> Dismissed @irfansharif from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-:-MwgbTnb7PLzPNYXbMzJ:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @irfansharif)

___
*[pkg/spanconfig/systemtarget.go, line 158 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl20-r3-158:-Mwg7wE_33VprXuK0ml2:b-idrbf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/systemtarget.go#L158)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why do we need both this and `st.targetsEntireKeyspace` above?
</blockquote></details>

Oops, meant to delete the thing above. 
___
*[pkg/spanconfig/systemtarget.go, line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl20-r3-167:-Mwgb1FREbFJxHmwwThd:bsia8i0) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/systemtarget.go#L167)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This panic feels a bit misplaced/confusing. Not apparent that we're doing any writes here in this API.
</blockquote></details>

Switched to returning an error and swallowing it at the caller -- not sure if it's much better, but maybe it's okay considering we never expect this to happen. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r3-42:-MwgU9daAtH31EHPkb13:bf7kt0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L42)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Instead of concatenating and making use of _'s, what if this syntax captured exactly what was being appended? 
```
conf=D+X
```
</blockquote></details>

I'm not sure how we'd make the syntax capture what's being appended at this layer. For context, the "combining" happens at the level of the store -- the subscriber is agnostic to which PTS field comes from which component (span config, system span config). 

Maybe all you're asking for is to switch `_CLUSTER` to something like `+X`? If that's the case, I've made the change. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r3-52:-MwgYKoi0JIfL5DCFpK3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L52)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Here too I'd prefer using single letters instead and showing combination of configs through "A+B+C".
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 57 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r3-57:-MwgYMQp1kZrGHtbx5Cn:b-cr3xo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L57)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is this right? I thought the system tenant targeting itself would would only capture its own keyspace without going into the tenant's. Wasn't that the whole point of separating out the EntireKeyspaceTarget into its own thing?
</blockquote></details>

Switched this to only include system tenant keys. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f8cI6nln0PwElN-r3-49:-MwgYbY7ETnx6Q8yhH_2:b-rih8wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go#L49)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why the _? Drop the cluster verbiage. The text here is just a shorthand for a config, using any letter in the alphabet instead would avoid confusing readers that these words/underscore have special meaning. Mind doing that?
</blockquote></details>

Done; though in the tests we use things like `+X` to visually differentiate the output.
___
*[pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go, line 248 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f8cI6nln0PwElN-r3-248:-MwgYurH8FTU3oOSfY_S:b-9wuh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/datadriven_test.go#L248)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why did we need this?
</blockquote></details>

Switched this around a bit; we need some special pretty printing for spans not constructed using our re. I moved it to `PrintSpan` though. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go, line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f8cI6nln0PwElL-r3-98:-Mwg73av7yGQCjIUGM9B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da01fdba4563dff59b540b0c0f14bdc103d3b290/pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go#L98)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Drop the "cluster" verbiage here and throughout.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go:
```diff
@@ -248,15 +249,25 @@ func (s *KVSubscriber) handlePartialUpdate(
+				var err error
+				sp, err = target.GetSystemTarget().KeyspaceTargeted()
+				if err != nil {
+					// Swallow the error here, even though we never expect this to happen.
```
This is still a bit odd, and tightly coupled to internals of KeyspaceTargeted. I think you want a check at this level that the SystemTargetType isn't the odd one you're not expecting and fatal-ing if it is. With that, KeyspaceTargeted could `panic("unimplemented")` or `panic("not applicable")`. I should've been clearer earlier, I was just thrown off by the "read only target" message in the panic.

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go:
```diff
@@ -86,6 +77,84 @@ func TestSpanConfigDecoder(t *testing.T) {
+
+		require.Equal(t, conf, got.Config)
+		require.True(t, got.Target.IsSystemTarget())
+		require.Equal(t, systemTarget, got.Target.GetSystemTarget(), "i:%d", i)
```
Is `"i:%d", i)` debugging detritus?

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+# The system span config we set should still apply.
+store-reader key=g
+----
+conf=MISSING+X
```
[nit] s/MISSING/FALLBACK

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+----
+conf=MISSING+X
+
+# Update span configs that target tenant keyspaces and ensure handlers are
```
Move everything below this line into a separate file; the spans don't overlap with the special [alphabet, alphabet) style spans listed above, except for `{source=1, target=1}:` which could stay in this file.

irfansharif commented on pkg/spanconfig/spanconfigtestutils/utils.go:
```diff
@@ -247,7 +247,21 @@ func ParseStoreApplyArguments(t *testing.T, input string) (updates []spanconfig.
+		sp.Key.String(),
+		sp.EndKey.String(),
+	}
+	for i := range s {
```
This is breaking symmetry with ParseSpan, worth at least updating the comment above.  Typo below "qe".

irfansharif commented on pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs:
```diff
@@ -0,0 +1,127 @@
+
+store-reader key=a
+----
+conf=A+Y
```
Just looked into the earlier PR that took character tags and int-casted them into protection policies. To me that at least smells weird, I don't know. If all we cared about was unique tags, it's fine to tag them using integers. When printing the configs, instead of this string concat (we're having to use tags prefixed with "+X", which feels odd), we could simply print a list of protection policies in PrintSpanConfig with "+", iterating through each PTS.

irfansharif reviewed (approved):
I'll rubber stamp because it's a testing only PR, but I think we should make the test changes I'm suggesting.

arulajmani reviewed (commented):
I reverted the patch I had to fix the bug in how we invoke our handlers; it turns out there's more we need to do there and I'll pull it out into its own PR. 

> Dismissed @irfansharif from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-:-MwhX-OQEKBWht8SiNNq:bjoaqmz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @irfansharif)

___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl1y-r5-263:-MwhMuik7ChxSIz3YMnm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L263)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This is still a bit odd, and tightly coupled to internals of KeyspaceTargeted. I think you want a check at this level that the SystemTargetType isn't the odd one you're not expecting and fatal-ing if it is. With that, KeyspaceTargeted could `panic("unimplemented")` or `panic("not applicable")`. I should've been clearer earlier, I was just thrown off by the "read only target" message in the panic.
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigtestutils/utils.go, line 254 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-MwgbU1orRETzMUDAPUp-r5-254:-MwhNvV_9qGZF49n7b6u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigtestutils/utils.go#L254)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

This is breaking symmetry with ParseSpan, worth at least updating the comment above.  Typo below "qe".
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r5-47:-MwhTh_M3WgNkHAfr0VZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L47)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] s/MISSING/FALLBACK
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r5-49:-MwhSspt0Xi7f-6ThNCI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L49)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Move everything below this line into a separate file; the spans don't overlap with the special [alphabet, alphabet) style spans listed above, except for `{source=1, target=1}:` which could stay in this file.
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 85 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r5-85:-MwhVJeYCExqTPgzA8CB:bvwqvjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L85)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Just looked into the earlier PR that took character tags and int-casted them into protection policies. To me that at least smells weird, I don't know. If all we cared about was unique tags, it's fine to tag them using integers. When printing the configs, instead of this string concat (we're having to use tags prefixed with "+X", which feels odd), we could simply print a list of protection policies in PrintSpanConfig with "+", iterating through each PTS.
</blockquote></details>

The construction here allows us to store words, not just characters to tag a config. I quite like how we can use this to do neat tricks like call a tagged config `FALLBACK`; I'm ambivalent to to this "+X" / int cast given that. I won't hold up the merging this PR over this, considering it's a nitty detail, but happy to retroactively fix this representation. 
___
*[pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go, line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f8cI6nln0PwElL-r5-154:-MwhUeUq2sNATnTayszS:b-p39rju) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/spanconfigdecoder_test.go#L154)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Is `"i:%d", i)` debugging detritus?
</blockquote></details>

Yeah, removed. 


<!-- Sent from Reviewable.io -->

arulajmani commented:
Also added a test for `KeyspaceTargeted` like we talked about offline.

arulajmani committed:
spanconfigtestutils: change how we tag configs

Previously, we could tag configs using arbitrary words. For tests that
used both span configs and system span configs, we'd end up needing to
put visual delimiters when setting up the test ourselves. This patch
makes it such that we can only tag configs using a single character.
We then render delimiters ourselves when printing "combined" configs.

For readability, we still allow setting a config called "FALLBACK".
This was the only reason words, as opposed to chars, were used to tag
configs.

Release note: None

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @irfansharif)

___
*[pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs, line 85 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76942#-Mwc28f9wrMpayR3Pl21-r5-85:-Mwhi7kV2t8gq54ECtRB:b-rze962) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83e5b32b1746c36a53744523e228f1e3f0c1238/pkg/spanconfig/spanconfigkvsubscriber/testdata/system_span_configs#L85)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The construction here allows us to store words, not just characters to tag a config. I quite like how we can use this to do neat tricks like call a tagged config `FALLBACK`; I'm ambivalent to to this "+X" / int cast given that. I won't hold up the merging this PR over this, considering it's a nitty detail, but happy to retroactively fix this representation. 
</blockquote></details>

Tacked on 4th commit on here to get around using the "+X" style tags; we now instead require tags to be a single character, or, "FALLBACK" which we special case for some readability benefits. I like this approach over the switching to integers to tag configs as that'd require unnecessary test case churn. Or maybe I'm just lazy and too used to seeing letter tags :P  


<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTR! 

bors r+
